### ISSUE TITLE
BUG:  HDR2SDR Color Issue When Cropping

### ISSUE THREAD

**Issue by parasupaman** [rigaya/QSVEnc#57]
HDR2SDR appears to be working great on my Tiger Lake CPU except for the following cropping issue.

With **_no cropping_**, notice that the HDR2SDR conversion works great:
```
D:\MKV\HDR>QSVEncC64.exe --avhw --fps 24000/1001 --codec hevc --sar 1:1 --qp-offset 2:6:8 --async-depth 2 --no-deblock --ref 5 --bframes 7 --b-pyramid --gop-len 240 --weightp --weightb --fade-detect --vpp-colorspace hdr2sdr=reinhard --cqp 18:19:21 --quality best --input-file Replicas.hevc --output-file Replicas.hevc.mkv
--------------------------------------------------------------------------------
Replicas.hevc.mkv
--------------------------------------------------------------------------------
cop3.DirectBiasAdjustment value changed off -> auto by driver
cop3.GlobalMotionBiasAdjustment value changed off -> auto by driver
QSVEncC (x64) 5.02 (r2308) by rigaya, May 16 2021 11:09:08 (VC 1928/Win/avx2)
OS             Windows 10 x64 (19042) [UTF-8]
CPU Info       11th Gen Intel Core i7-1165G7 @ 2.80GHz [TB: 4.10GHz] (4C/8T) <Tigerlake>
GPU Info       Intel Iris(R) Xe Graphics (96EU) 100-1300MHz [28W] (27.20.100.9466)
Media SDK      QuickSyncVideo (hardware encoder) PG, 1st GPU, API v1.34
Async Depth    2 frames
Buffer Memory  d3d11, 30 work buffer
Input Info     avqsv: H.265/HEVC, 3840x2160, 24000/1001 fps
VPP            cspconv(p010 -> yv12)
               colorspace: cspconv(yv12 -> yuv444)
                                          matrix:bt2020nc->GBR
                                          transfer:smpte2084->linear
                                          hdr2sdr(reinhard): source_peak=1000.00 ldr_nits=100.00
                                            contrast 0.50, peak 1.00
                                            desat base 0.18, strength 0.75, exp 1.50
                                          prim:bt2020->bt709
                                          transfer:linear->bt709
                                          matrix:GBR->bt709
               cspconv(yuv444 -> nv12)
AVSync         cfr
Output         HEVC(yuv420) main @ Level 5 (high tier)
               3840x2160p 1:1 23.976fps (24000/1001fps)
               avwriter: hevc => matroska
Target usage   1 - best
Encode Mode    Constant QP (CQP)
CQP Value      I:18  P:19  B:21
QP Limit       min: 1, max: 51
Trellis        Auto
Ref frames     5 frames
Bframes        7 frames, B-pyramid: on
Max GOP Length 240 frames
Ext. Features  No-Deblock WeightP WeightB FadeDetect QPOffset tskip ctu:64 sao:all
```
![HDR-Reinhard-No Crop](https://user-images.githubusercontent.com/77555688/118908214-4f1b7600-b964-11eb-81d1-1bacc5972e0a.jpg)

However, notice that the **colors now appear skewed** when _**cropping**_ the image:
```
D:\MKV\HDR>QSVEncC64.exe --avhw --fps 24000/1001 --codec hevc --sar 1:1 --qp-offset 2:6:8 --async-depth 2 --no-deblock --ref 5 --bframes 7 --b-pyramid --gop-len 240 --weightp --weightb --fade-detect --vpp-colorspace hdr2sdr=reinhard --crop 0,276,0,276 --cqp 18:19:21 --quality best --input-file Replicas.hevc --output-file Replicas.hevc.mkv
--------------------------------------------------------------------------------
Replicas.hevc.mkv
--------------------------------------------------------------------------------
cop3.DirectBiasAdjustment value changed off -> auto by driver
cop3.GlobalMotionBiasAdjustment value changed off -> auto by driver
QSVEncC (x64) 5.02 (r2308) by rigaya, May 16 2021 11:09:08 (VC 1928/Win/avx2)
OS             Windows 10 x64 (19042) [UTF-8]
CPU Info       11th Gen Intel Core i7-1165G7 @ 2.80GHz [TB: 4.10GHz] (4C/8T) <Tigerlake>
GPU Info       Intel Iris(R) Xe Graphics (96EU) 100-1300MHz [28W] (27.20.100.9466)
Media SDK      QuickSyncVideo (hardware encoder) PG, 1st GPU, API v1.34
Async Depth    2 frames
Buffer Memory  d3d11, 30 work buffer
Input Info     avqsv: H.265/HEVC, 3840x2160, 24000/1001 fps
VPP            crop: 0,276,0,276/cspconv(p010 -> yv12)
               colorspace: cspconv(yv12 -> yuv444)
                                          matrix:bt2020nc->GBR
                                          transfer:smpte2084->linear
                                          hdr2sdr(reinhard): source_peak=1000.00 ldr_nits=100.00
                                            contrast 0.50, peak 1.00
                                            desat base 0.18, strength 0.75, exp 1.50
                                          prim:bt2020->bt709
                                          transfer:linear->bt709
                                          matrix:GBR->bt709
               cspconv(yuv444 -> nv12)
AVSync         cfr
Output         HEVC(yuv420) main @ Level 5 (high tier)
               3840x1608p 1:1 23.976fps (24000/1001fps)
               avwriter: hevc => matroska
Target usage   1 - best
Encode Mode    Constant QP (CQP)
CQP Value      I:18  P:19  B:21
QP Limit       min: 1, max: 51
Trellis        Auto
Ref frames     5 frames
Bframes        7 frames, B-pyramid: on
Max GOP Length 240 frames
Ext. Features  No-Deblock WeightP WeightB FadeDetect QPOffset tskip ctu:64 sao:all
```
![HDR-Reinhard-Cropped](https://user-images.githubusercontent.com/77555688/118908568-00baa700-b965-11eb-9c17-8a440ed17e60.jpg)

The same issue appears to occur **regardless** of the HDR2SDR type (eg. Hable, Mobius, Reinhard, bt2390 etc.).

Also, excellent work with the HDR2SDR feature!  I have been really looking forward to GPU tone-mapping.  I would love to continue working (medical pathology programmer), while I shift 4K tone mapping encodes into the background (from CPU to GPU) - goodbye CPU tone-mapping.

Hope that the fix is not too difficult to troubleshoot.  Cheers!

**Comment by parasupaman** [comment]
**_Color skew issue still exists in version 5.03_** when using **hardware decoding (--avhw)**.  Issue _**does not exist using software decoding (--avsw)**_.  However, --avsw decoding introduces stutters to the video playback (as detailed in issue https://github.com/rigaya/QSVEnc/issues/58)

**Comment by ukmark62** [comment]
Unfortunately, I cannot get --avhw to work with HDR files on Ice Lake. The encoder reverts to --avsw, where, as you point out, colours are fine but playback is juddery. With SDR files, --avhw works every time.

**Comment by parasupaman** [comment]
> Unfortunately, I cannot get --avhw to work with HDR files on Ice Lake. The encoder reverts to --avsw, where, as you point out, colours are fine but playback is juddery. With SDR files, --avhw works every time.

Interesting.  I am guessing that hardware decoding of HDR content is only supported by Tiger Lake.  GPU HDR tone-mapping is a big part of the reason I wanted to upgrade to Tiger Lake this year (from my aging Kaby Lake based machine).  Has been a brilliant upgrade.

**Comment by ukmark62** [comment]
I very rarely use HDR content. I'm very happy with the SDR side of Ice Lake. I also had old Kaby Lake PC. I find the CQP hybrid mode quality on Ice Lake to be really good. In my subjective tests, it is as good as x265 10-bit medium preset.  Maybe in a few years, I'll upgrade again. I can get HDR to SDR working with FFMPEG and can get around 23fps for a 1080p encode using QSV HEVC 10-bit. 
The only thing that I can't get with FFMPEG is the "qp-offset of 2:6:8" which allows a higher quality CQP setting and still keeps file sizes small. 
Hopefully, if the stuttering bug is fixed with HDR2SDR then I'll be able to make use of that with --avsw encoding, even if HW encoding is still not available for Ice Lake.

**Comment by parasupaman** [comment]
> I very rarely use HDR content. I'm very happy with the SDR side of Ice Lake. I also had old Kaby Lake PC. I find the CQP hybrid mode quality on Ice Lake to be really good. In my subjective tests, it is as good as x265 10-bit CRF medium preset. Maybe in a few years, I'll upgrade again. I can get HDR to SDR working with FFMPEG and can get around 23fps for a 1080p encode using QSV HEVC 10-bit.
> The only thing that I can't get with FFMPEG is the "qp-offset" of 2:6:8 which allows a higher quality CQP setting and still keeps file sizes small.
> Hopefully, if the stuttering bug is fixed with HDR2SDR then I'll be able to make use of that with --avsw encoding, even if HW encoding is still not available for Ice Lake.

I agree.  CQP hybrid quality mode is excellent (especially with the "qp-offset 2:6:8" tip that I got from one of your previous posts - thanks!).  And yeah, probably not worth upgrading anytime soon, since you are already on Ice Lake.

I also mainly encode SDR.  However, am keen to start 4K HDR-to-SDR tone-mapped encodes.

**Comment by rigaya** [comment]
I've made further fix with the crop issue [QSVEnc 5.04](https://github.com/rigaya/QSVEnc/releases/tag/5.04) when used with --avhw. Hope this helps.

**Comment by parasupaman** [comment]
> I've made further fix with the crop issue [QSVEnc 5.04](https://github.com/rigaya/QSVEnc/releases/tag/5.04) when used with --avhw. Hope this helps.

_HDR2SDR color shift_ issue when cropping is **now fixed** (also for **--avhw**).  Great job!  Will now enjoy some great hardware tone-mapping on Tiger Lake :)  Thanks very much!
